pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java

```diff
@@ -145,6 +137,32 @@ public void init(TableConfig tableConfig, Schema schema, TableDataManager tableD
     }
   }
 
+  /**
+   * Reuses a PartialUpsertHandler if there is no custom row merger otherwise creates a new instance to avoid
+   * concurrent modification of state(LazyRow) between segments of the table
+   * @return partial upsert handler to merge previous and new row
+   */
+    protected PartialUpsertHandler getPartialUpsertHandler() {
+        UpsertConfig upsertConfig = _tableConfig.getUpsertConfig();
+        if (upsertConfig.getMode() == UpsertConfig.Mode.PARTIAL) {
+        Map<String, UpsertConfig.Strategy> partialUpsertStrategies = upsertConfig.getPartialUpsertStrategies();
+        Preconditions.checkArgument(partialUpsertStrategies != null,
+                "Partial-upsert strategies must be configured for partial-upsert enabled table: %s",
+                _tableNameWithType);
+        if (upsertConfig.getRowMergerCustomImplementation() != null) {
+          // In case of custom merger, the PUH is dependent on LazyRow object to be reused. LazyRow is stateful and
+          // cause concurrent modification issues, hence a new PUH is created per partition
+          return new PartialUpsertHandler(_schema, upsertConfig, _comparisonColumns);
```

Comment:
Got it. currently the partial upsert handler are initialized in the upsertTableManager and shared by partitionUpsertManager. After lazyRow passed in the partialUpsertHandler it become stateful, and multiple partitions can access the partial upsert handler. this leads to concurrent modification issue. So we have to avoid sharing the partial upsert handler among partitions. 

@Jackie-Jiang do you see any improvement on this approach?